### PR TITLE
LPS-93540 Connect to redux devtools

### PR DESCRIPTION
**Pull request by p2kmgcl**

### CODE REVIEW

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js [hunk 1]
```diff
@@ -97,6 +97,21 @@ class Store extends State {
...
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	disposed() {
+		if (this._devTools) {
+			this._devTools.disconnect();
+		}
```
Comment (wincent) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js]: I'd wrap this in the `process.env.NODE_ENV` check as will so that it can in theory be entirely stripped out in production builds.

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js [hunk 2]
```diff
@@ -117,24 +132,37 @@ class Store extends State {
 					)
 				),
 				Promise.resolve(this._state)
-			)
-				.then(
-					nextState => {
+			).then(
+				nextState => {
+					if (this._state !== nextState) {
```
Comment (wincent) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js]: Seems like this should be pulled out into an independent commit.

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js [hunk 3]
```diff
@@ -117,24 +132,37 @@ class Store extends State {
...
-								);
-							}
-						);
+						if (this._devTools) {
+							this._devTools.send(
+								Object.assign(
+									{
+										type: actionType
+									},
+									payload
```
Comment (wincent) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js]: Is `payload` immutable? Just because `Object.assign` only does a shallow copy, and I am not sure what happens (could happen) to it after you call `send()`.

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js [hunk 4]
```diff
@@ -117,24 +132,37 @@ class Store extends State {
...
-						return new Promise(
-							resolve => {
-								requestAnimationFrame(
-									() => {
-										resolve(this);
-									}
-								);
-							}
-						);
+						if (this._devTools) {
```
Comment (wincent) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js]: Same thing here about the env check.

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js (hunk 3 above)
Comment (p2kmgcl) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js]: Well, it's not a frozen or so, but it should not be mutated anywhere. Technically it only logs the given payload to the devtools history. Moreover, our payloads are (I think always) plain objects without nested properties.

Should I still protect us from mutations?

Comment (wincent) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js]: The reason I brought it up is that the use of `Object.assign` makes it look like you're trying to make a copy, but it is only a shallow copy. That begs the question, are you worried about mutation? And if so, why only a shallow copy?

Personally, rather than making it a deep copy, I'd actually just drop the `Object.assign`; you are already creating a brand new object anyway with that `{type: actionType}` literal, so you may as well just use that. eg.

```javascript
this._devTools.send({
    type: actionType,
    payload,
});
```

of course, I have no idea if we can ES5 object short notation in this project. What transforms do we have? If in doubt:

```javascript
this._devTools.send({
    type: actionType,
    payload: payload,
});
```

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js [hunk 5]
```diff
@@ -1,6 +1,11 @@
 import State, {Config} from 'metal-state';
 import {DEFAULT_INITIAL_STATE} from './state.es';
 
+/**
+ * ID of the development devTool that may be connected to store
```
Comment (wincent) [inline modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js]: This would be a great place to add a link to the upstream project explaining what this mysterious thing is (I know you included it in the commit message, but it would be good to have it here).

File: modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js (hunk 3 above)
Comment (p2kmgcl) [reply modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/js/store/store.es.js]: I was trying to produce a more redux-alike action object, but your proposal looks much more clean, so I'm taking that. Thanks!

FYI: we do have object short notation :D
